source/upyun/upyun.go

```diff
@@ -0,0 +1,143 @@
+// +-------------------------------------------------------------------------
+// | Copyright (C) 2016 Yunify, Inc.
+// +-------------------------------------------------------------------------
+// | Licensed under the Apache License, Version 2.0 (the "License");
+// | you may not use this work except in compliance with the License.
+// | You may obtain a copy of the License in the LICENSE file, or at:
+// |
+// | http://www.apache.org/licenses/LICENSE-2.0
+// |
+// | Unless required by applicable law or agreed to in writing, software
+// | distributed under the License is distributed on an "AS IS" BASIS,
+// | WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// | See the License for the specific language governing permissions and
+// | limitations under the License.
+// +-------------------------------------------------------------------------
+
+package upyun
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	log "github.com/frostyplanet/logrus"
+	"github.com/upyun/go-sdk/upyun"
+	"github.com/yunify/qscamel/record"
+)
+
+const upyunEndpoint = ".upaiyun.com"
+
+// SourceUpyun is upyun bucket type source
+type SourceUpyun struct {
+	BucketURL  string
+	bucketName string
+	up         *upyun.UpYun
+
+	lastModifiedTimes map[string]time.Time
+	objsChan          chan *upyun.FileInfo
+	marker            string
+}
+
+var errMissCredential = errors.New("Miss accessKeyID or secretAccessKey for upyun")
+
+// NewSourceUpyun creates an instance of SourceUpyun
+func NewSourceUpyun(bucketName, zone, accessKeyID, secretAccessKey string) (*SourceUpyun, error) {
+	if accessKeyID == "" || secretAccessKey == "" {
+		return &SourceUpyun{}, errMissCredential
+	}
+
+	up := upyun.NewUpYun(&upyun.UpYunConfig{
+		Bucket:   bucketName,
+		Operator: accessKeyID,
+		Password: secretAccessKey,
+	})
+
+	fmt.Println("http://" + bucketName + "." + zone + upyunEndpoint)
+
+	return &SourceUpyun{
+		BucketURL:         "http://" + bucketName + "." + zone + upyunEndpoint,
+		bucketName:        bucketName,
+		up:                up,
+		lastModifiedTimes: make(map[string]time.Time),
+		marker:            "",
+	}, nil
+}
+
+// GetSourceSites implements MigrateSource.GetSourceSites
+func (source *SourceUpyun) GetSourceSites(
+	threadNum int, logger *log.Logger, recorder *record.Recorder,
+) (sourceSites []string, objectNames []string, skipped []string, done bool, err error) {
+	sourceSites, objectNames, skipped = []string{}, []string{}, []string{}
+	err = nil
+
+	if source.marker == "" {
+		source.objsChan = make(chan *upyun.FileInfo, 10)
+		// upyun golang sdk does not suport to get part of all the objects to do advanced treatment.
+		// Inside the "List" interface, it Summarize all objects got from diffent http resp
+		// So we store all the objects only at first time.
+		// Then we use the sotred objects to do partial traversal regarding the number of concurrent tasks
```

Comment:
source.Bucket is sth like “http://XXXXX”, use  path.Join() will  convert it into “http:/XXXXXX”,  maybe we should use strings.Join to keep “//”.